TypeSystem/flow-analysis/reachability_A18_t02.dart

```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion - Shortcut and: If `N` is a shortcut "and" expression of the form
+/// `E1 && E2`, then:
+/// - Let `before(E1) = before(N)`.
+/// - Let `before(E2) = split(true(E1))`.
+/// - Let `true(N) = unsplit(true(E2))`.
+/// - Let `false(N) = merge(split(false(E1)), false(E2))`.
+///
+/// @description Checks that in an expression of the form `E1 && E2` if `E1` is
+/// `true` then `E2` is always executed.
+/// @author sgrekhov22@gmail.com
+
+main() {
+  int i;
+  true && (i = 42) != 42;
+  i; // Definitely assigned
```

Comment:
I can see (in DartPad) that the flow analysis does conclude that `i = 42` is definitely being executed. It is not so obvious how this would be a necessary consequence of the rules in the `@assertion`.

But `false(E1)` is unreachable, so `split(false(E1))` is unreachable, so the `merge(...)` would be the same as `false(E2)`. This implies that `E2` has been evaluated if we're in the false continuation of the expression `E1 && E2`.

Similarly, `before(E2)` is `before(N)` (because `true(E1)` is `before(E1)`), so if `E1 && E2` is reached then `E2` is also reached.

OK!